src/main/java/com/google/devtools/build/lib/skylarkbuildapi/SkylarkRuleFunctionsApi.java

```diff
@@ -78,198 +78,203 @@
+            generic1 = String.class,
+            defaultValue = "[]",
+            doc =
+                "List of names of configuration fragments that the rule requires "
+                    + "in target configuration."),
+        @Param(
+            name = "host_fragments",
+            type = SkylarkList.class,
+            legacyNamed = true,
+            generic1 = String.class,
+            defaultValue = "[]",
+            doc =
+                "List of names of configuration fragments that the rule requires "
+                    + "in host configuration."),
+        @Param(
+            name = "_skylark_testable",
+            type = Boolean.class,
+            legacyNamed = true,
+            defaultValue = "False",
+            doc =
+                "<i>(Experimental)</i><br/><br/>"
+                    + "If true, this rule will expose its actions for inspection by rules that "
+                    + "depend on it via an <a href=\"globals.html#Actions\">Actions</a> "
+                    + "provider. The provider is also available to the rule itself by calling "
+                    + "<a href=\"ctx.html#created_actions\">ctx.created_actions()</a>."
+                    + "<br/><br/>"
+                    + "This should only be used for testing the analysis-time behavior of "
+                    + "Skylark rules. This flag may be removed in the future."),
+        @Param(
+            name = "toolchains",
+            type = SkylarkList.class,
+            legacyNamed = true,
+            generic1 = String.class,
+            defaultValue = "[]",
+            doc =
+                "<i>(Experimental)</i><br/><br/>"
+                    + "If set, the set of toolchains this rule requires. Toolchains will be "
+                    + "found by checking the current platform, and provided to the rule "
+                    + "implementation via <code>ctx.toolchain</code>."),
+        @Param(
+            name = "doc",
+            type = String.class,
+            legacyNamed = true,
+            defaultValue = "''",
+            doc =
+                "A description of the rule that can be extracted by documentation generating tools."),
+        @Param(
+            name = "provides",
+            type = SkylarkList.class,
+            named = true,
+            positional = false,
+            defaultValue = "[]",
+            doc =
+                "A list of providers this rule is guaranteed to provide. "
+                    + "It is an error if a provider is listed here and the rule "
+                    + "implementation function does not return it."),
+        @Param(
+            name = "execution_platform_constraints_allowed",
+            type = Boolean.class,
+            named = true,
+            positional = false,
+            defaultValue = "False",
+            doc =
+                "Whether targets of this rule can set exec_compatible_with to add extra constraints on the execution platform."),
```

Comment:
Yes: it's a list of labels, which must exist.
No: They can't be resolved as Targets, because execution constraints have to be checked before dependencies are resolved (because the selected toolchains are dependencies, and we need the execution constraints in order to select the proper toolchain).

The user can probably access ctx.attr.exec_compatible_with, I'm just not sure if there's ever a use for that.

Reply:
With or without a use for `ctx.attr.exec_compatible_with`, if it exists it should be documented, and the easiest way to do that is to refer to it as an "attribute" here.

If setting this bool causes it to be illegal for the user to define their own "exec_compatible_with" attribute in the `attrs` dict, we should also say so here, e.g.:

    If true, a special attribute named <code>exec_compatible_with</code> of label-list type is added, which must not already exist in `attrs`. ...

Reply:
(backticks to `<code>` and `</code>`)